### PR TITLE
support KubeletConfiguration and flag passthrough in NodeConfig 

### PR DESCRIPTION
**Pull request by ndbaker1**
**Issue #, if available:**

**Description of changes:**

supporting [KubeletConfiguration](https://pkg.go.dev/k8s.io/kubelet/config/v1beta1#KubeletConfiguration) parameters through the `NodeConfig` as pass through as well as the kubelet cli flags.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Testing Done**

added e2e tests to cover the additional configuration surface area

*[See this guide for recommended testing for PRs.](../doc/CONTRIBUTING.md#testing-changes) Some tests may not apply. Completing tests and providing additional validation steps are not required, but it is recommended and may reduce review time and time to merge.*

### CODE REVIEW

**Comment by cartermckinnon** [comment]
Chatted offline, I think we have to find a more generic way to accept kubelet config customizations. Clients like Karpenter want to pass a defined subset of the options, but clients like `eksctl` want to set arbitrary options: https://github.com/eksctl-io/eksctl/blob/9575570b554610129382d0a181645a3806cea98f/pkg/apis/eksctl.io/v1alpha5/types.go#L1247

We shouldn't need to make a code change here for every kubelet config option that every user wants to modify, it's toil for the maintainers and an artificial limitation for users.

I don't think we can do any meaningful validation of these options; when `nodeadm` runs, it's already too late to tell the user whether or not their kubelet config is valid. Kubelet will tell us itself. :)

**Comment by ellistarn** [comment]
> We shouldn't need to make a code change here for every kubelet config option that every user wants to modify, it's toil for the maintainers and an artificial limitation for users.

I'm curious to interrogate this assumption. I think that there's value in knowing how customers are using your API, and intentionally constraining things. I think we can pretty trivially fine examples of kubeletconfiguration that would never be used.

**Comment by cartermckinnon** [comment]
> knowing how customers are using your API

I think this data is valuable, but I think asking folks to cut an issue/PR here, get it merged, and wait for a release is a lot of friction just to permit them to set a field in the kubelet config file. If it were me, I'd just hack around it myself.

> I think we can pretty trivially fine examples of kubeletconfiguration that would never be used.

If the knob exists, someone wants (or perhaps less often, needs) to turn it. 😉

File: nodeadm/api/v1alpha1/nodeconfig_types.go
Comment (ndbaker1) [inline nodeadm/api/v1alpha1/nodeconfig_types.go:59]: Is this a premature addition of the `labels` and `taints` fields at the top level? they are technically configurable through `inline` config or `flags`, and i dont think we need to do any special validation with them.

Many users might expect `labels` and `taints` coming from the previous bootstrap.sh, but there's not a need to uphold that image

Comment (cartermckinnon) [reply nodeadm/api/v1alpha1/nodeconfig_types.go:59]: Managed nodegroups has passed these with the `--kubelet-extra-args` flag, so I think it's fine to let orchestrators use the generic kubelet options directly.

**Comment by mmerkes** [comment]
> I think that there's value in knowing how customers are using your API, and intentionally constraining things

For the Karpenter use case, I think this makes sense because there's some inherent expectations that you can make. However, in the case of these AMIs as a whole, the use cases are pretty vast. I don't think that we can realistically decide what parameters are and are not useful to customers without just allowing a majority of them to be set regardless, and this code will be significantly less maintainable because any new (or unsupported) config that needs to be added will require a PR. Given, may be a minor PR, but will still need to be cut and then released. Some people will just hope to hack their user data with `jq` to get it done, which is less desirable. There will be undoubtedly be niche use cases where very few actually need to support something, but it may be very important to them.

I like the idea of making it easy for customers to configure the fields they want and easily adopt new config options.

File: nodeadm/api/v1alpha1/nodeconfig_types.go
Comment (ndbaker1) [reply nodeadm/api/v1alpha1/nodeconfig_types.go:59]: ill pull back on these for now, since an operator can still fully configure kubelet without them and the contracts for these via `--node-labels`, `--register-with-taints` isn't going to break anytime soon

File: nodeadm/internal/api/net.go
Comment (cartermckinnon) [inline nodeadm/internal/api/net.go:43]: What's the rationale for this? Seems like it should be working with the internal API types

File: nodeadm/internal/api/types.go
Comment (cartermckinnon) [inline nodeadm/internal/api/types.go:54]: Ah, you'll have to add this back, the internal API should be separate from the public API types. The internal/api/bridge package maps multiple supported versions of the public API onto a single internal representation, that should be used everywhere. That package is the only place under internal/ where the public API types should be imported

File: nodeadm/internal/kubelet/environment.go
Comment (cartermckinnon) [inline nodeadm/internal/kubelet/environment.go:32]: Does the value of this environment variable need to be quoted?

i.e.
```
KUBELET_ARGS=--foo --bar
```
or
```
KUBELET_ARGS='--foo --bar'
```

File: nodeadm/test/e2e/cases/kubeconfig/config.yaml
Comment (cartermckinnon) [inline nodeadm/test/e2e/cases/kubeconfig/config.yaml:5]: nit: this isn't necessary, right? we could omit this from all the test assets going forward

File: nodeadm/test/e2e/cases/kubelet-flags/config.yaml
Comment (cartermckinnon) [inline nodeadm/test/e2e/cases/kubelet-flags/config.yaml:14]: This usage looks like of odd. Why not use the container args approach:
```
flags:
  - -v=7
  - --cloud-provider=external
  - --image-credential-provider-config=/foo/bar
```

**Review by cartermckinnon — CHANGES_REQUESTED** [review]

File: nodeadm/internal/api/net.go
Comment (ndbaker1) [reply nodeadm/internal/api/net.go:43]: i wanted to work with the external types as much as possible to avoid having too an excess of duplicated types, but wasn't sure if it was a good idea to move this to the api package or just swap the signature like so

File: nodeadm/internal/kubelet/environment.go
Comment (ndbaker1) [reply nodeadm/internal/kubelet/environment.go:32]: didnt see any hard requirements for it to be quoted, let me triple check

File: nodeadm/internal/api/types.go
Comment (ndbaker1) [reply nodeadm/internal/api/types.go:54]: oh gotcha, i saw other projects using public types in their impls and misinterpreted this, i can quickly flip them all back

File: nodeadm/test/e2e/cases/kubelet-flags/config.yaml
Comment (ndbaker1) [reply nodeadm/test/e2e/cases/kubelet-flags/config.yaml:14]: The only real difference is that you are going to see all of the argument that nodeadm writes as well as the user provided arguments, when with the map you are actually overriding the value in-place; if users are more familiar with this approach then it makes sense to me. There shouldn't be a functional difference as long as the flags are appended to the end.

File: nodeadm/test/e2e/cases/kubeconfig/config.yaml
Comment (ndbaker1) [reply nodeadm/test/e2e/cases/kubeconfig/config.yaml:5]: true 👍

File: nodeadm/internal/kubelet/environment.go
Comment (ndbaker1) [reply nodeadm/internal/kubelet/environment.go:32]: going to fix this to double quotes per https://github.com/markstos/systemd/blob/master/man/systemd.exec.xml#L345-L363

File: nodeadm/test/e2e/cases/kubelet-flags/config.yaml
Comment (cartermckinnon) [reply nodeadm/test/e2e/cases/kubelet-flags/config.yaml:12]: The k8s components use spf13/pflag under the covers: https://github.com/kubernetes/component-base/blob/2065849162a4fdce3f03b9e255475ef0ef04390d/cli/flag/flags.go#L23

Which has the behavior that if a flag which only supports a single value is defined twice, the later value is used. So as long as what the user defines here follows our defaults (which is the case on AL2), it'll work as expected as far as overrides go.

File: nodeadm/test/e2e/cases/node-taints/run.sh
Comment (ellistarn) [inline nodeadm/test/e2e/cases/node-taints/run.sh:24]: Would be sweet if we could do 
```
cat <<EOF | kubectl init -f -
... inline file
EOF
```

File: nodeadm/test/e2e/cases/kubelet-flags/config.yaml
Comment (ellistarn) [inline nodeadm/test/e2e/cases/kubelet-flags/config.yaml:9]: Does this cause NodeConfig to inject kubelet flags? What if they conflict with user defined flags?

Comment (ndbaker1) [reply nodeadm/test/e2e/cases/kubelet-flags/config.yaml:9]: [carter just answered this one](https://github.com/awslabs/amazon-eks-ami/pull/1593#discussion_r1467083198) while we determined the input style; we will generate flags in nodeadm and then user-provided flags can override any of these

File: nodeadm/test/e2e/cases/node-taints/run.sh
Comment (ndbaker1) [reply nodeadm/test/e2e/cases/node-taints/run.sh]: most users wont have a need to actually call `nodeadm` based on how its booted in the ami, but im a fan of the style

File: nodeadm/test/e2e/cases/kubelet-flags/config.yaml
Comment (ndbaker1) [reply nodeadm/test/e2e/cases/kubelet-flags/config.yaml]: done, makes sense to use the typical command args approach

File: nodeadm/api/v1alpha1/nodeconfig_types.go
Comment (ndbaker1) [reply nodeadm/api/v1alpha1/nodeconfig_types.go]: im going with `config` for the brevity 😬

**Review by cartermckinnon — APPROVED** [review]

**Pull request merged**
